packages/pyright-internal/src/analyzer/importResolver.ts

```diff
@@ -1310,14 +1324,112 @@ export class ImportResolver {
         }
     }
 
-    private _getStdlibTypeshedPath(execEnv: ExecutionEnvironment, importFailureInfo: string[]) {
-        return this._getTypeshedSubdirectory(/* isStdLib */ true, execEnv, importFailureInfo);
+    // Returns the directory for a module within the stdlib typeshed directory.
+    // If moduleDescriptor is provided, it is filtered based on the VERSIONS
+    // file in the typeshed stubs.
+    private _getStdlibTypeshedPath(
+        execEnv: ExecutionEnvironment,
+        importFailureInfo: string[],
+        moduleDescriptor?: ImportedModuleDescriptor
+    ) {
+        const subdirectory = this._getTypeshedSubdirectory(/* isStdLib */ true, execEnv, importFailureInfo);
+        if (
+            subdirectory &&
+            moduleDescriptor &&
+            !this._isStdlibTypeshedStubValidForVersion(moduleDescriptor, execEnv, importFailureInfo)
+        ) {
+            return undefined;
+        }
+
+        return subdirectory;
     }
 
     private _getThirdPartyTypeshedPath(execEnv: ExecutionEnvironment, importFailureInfo: string[]) {
         return this._getTypeshedSubdirectory(/* isStdLib */ false, execEnv, importFailureInfo);
     }
 
+    private _isStdlibTypeshedStubValidForVersion(
+        moduleDescriptor: ImportedModuleDescriptor,
+        execEnv: ExecutionEnvironment,
+        importFailureInfo: string[]
+    ) {
+        if (!this._cachedTypeshedStdLibModuleVersions) {
+            this._cachedTypeshedStdLibModuleVersions = this._readTypeshedStdLibVersions(execEnv, importFailureInfo);
+        }
+
+        const versionRange = this._cachedTypeshedStdLibModuleVersions!.get(moduleDescriptor.nameParts[0]);
```

Comment:
Nit: Is `!` necessary here? Typescript should have narrowed this due to the guard above.

Reply:
You're correct. Removed.